### ISSUE TITLE
[Flaky Test] Test E2E Gardener Suite - Shoot Tests Create, Update, Delete

### ISSUE THREAD

**Issue by timuthy** [gardener/gardener#7530]
**How to categorize this issue?**

/area testing
/kind flake

**Which test(s)/suite(s) are flaking**:

Suite: `Test E2E Gardener Suite`
Test: `Shoot Tests Create, Update, Delete` 

**CI link**:
https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7518/pull-gardener-e2e-kind/1627712984006529024

**Reason for failure**:
Shoot reconciliation is not considered successful:

```
 {"level":"info","ts":"2023-02-20T17:36:35.908Z","logger":"test","msg":"Shoot is not yet reconciled","shoot":"garden-local/e2e-default","reason":"condition type SystemComponentsHealthy is not true yet, had message Deployment \"kube-system/local-path-provisioner\" is unhealthy: condition \"Available\" has invalid status False (expected True) due to MinimumReplicasUnavailable: Deployment does not have minimum availability. with reason DeploymentUnhealthy"}
```

The `local-path-provisioner` was not able to resolve the Kube-Apiserver's host:

```
2023-02-20T17:32:08.814485735Z stderr F time="2023-02-20T17:32:08Z" level=fatal msg="Error starting daemon: Cannot start Provisioner: failed to get Kubernetes server version: Get \"https://api.e2e-default.local.internal.local.gardener.cloud:443/version?timeout=32s\": dial tcp: lookup api.e2e-default.local.internal.local.gardener.cloud: i/o timeout"
```

**Comment by acumino** [comment]
In this test case, we are creating three worker pools(each worker pool with one node). The test case is failing because `local-path-provisioner` is in `CrashLoopBackOff` because it can't reach `kube-apiserver`.

For `local-path-provisioner` to run successfully it needs to reach `kube-apiserver` and in our setup CoreDNS is supposed to resolve `kube-apiserver` domain `api.e2e-default.local.internal.local.gardener.cloud`. We run two replicas of coreDNS, so the request to resolve this domain can go to any of the coreDNS pods. 


### Scenario 1:  
If the name resolution request reaches the coreDNS pod running in the same node, it **succeeds**. From the TCP dump, we can observe request originates and ends at the same IP.
<details open>
<summary>TCP dump</summary>
<be>

```
14:12:41.122340 cali05a79585cfa In  IP 10.3.1.20.45632 > 10.4.0.10.53: 38162+ [1au] A? abc.com. (48)
14:12:41.122388 calif0ebd3497ab Out IP 10.3.1.20.45632 > 10.3.1.21.8053: UDP, length 48
14:12:41.130884 calif0ebd3497ab In  IP 10.3.1.21.8053 > 10.3.1.20.45632: UDP, length 140
14:12:41.130902 cali05a79585cfa Out IP 10.4.0.10.53 > 10.3.1.20.45632: 38162 4/0/1 A 18.155.99.34, A 18.155.99.95, A 18.155.99.128, A 18.155.99.3 (140)
```

</details>



### Scenario 2:  

If the name resolution request reaches the coreDNS pod running in another node it **fails**. From the TCP dump, we can observe the request originates from pod `10.3.1.20` but the final response never reaches the same pod, actually **_the request is not able to reach coreDNS_**,  because inter-node communication is not working. 
<details open>
<summary>TCP dump</summary>
<br>

```
14:13:57.971650 cali05a79585cfa In  IP 10.3.1.20.41869 > 10.4.0.10.53: 51232+ [1au] A? abc.com. (48)
14:13:57.971808 eth0  Out IP 10.3.1.20.41869 > 10.3.0.24.8053: UDP, length 48
14:14:02.970696 cali05a79585cfa In  IP 10.3.1.20.41869 > 10.4.0.10.53: 51232+ [1au] A? abc.com. (48)
14:14:02.970748 eth0  Out IP 10.3.1.20.41869 > 10.3.0.24.8053: UDP, length 48
14:14:03.352484 cali05a79585cfa In  ARP, Request who-has 169.254.1.1 tell 10.3.1.20, length 28
14:14:03.352572 cali05a79585cfa Out ARP, Reply 169.254.1.1 is-at ee:ee:ee:ee:ee:ee, length 28
14:14:07.970756 cali05a79585cfa In  IP 10.3.1.20.41869 > 10.4.0.10.53: 51232+ [1au] A? abc.com. (48)
14:14:07.970826 eth0  Out IP 10.3.1.20.41869 > 10.3.0.24.8053: UDP, length 48
```

</details>

<br>

<hr>
<br>

**TL:DR**
The test fails because `local-path-proivsioner` is in `CrashLoopBackoff` because it can't resolve `kube-apiserver` server host if coreDNS pod is running in a different node other than the `local-path-proivsioner` pod.
_As of present, inter-node communication is not supported between machine pods in `provider-local` setup._


Thanks, @seshachalam-yv and @ScheererJ for help in debugging.

**Comment by rfranzke** [comment]
Thanks for the detailed analysis! That's indeed quite helpful. Did you folks discuss next steps? What can we do now to solve or at least mitigate this problem?

**Comment by ScheererJ** [comment]
Well, if it only about name resolution and it can be ensured that every node gets a `coredns` replica [internal traffic policy](https://kubernetes.io/docs/concepts/services-networking/service-traffic-policy/) can be a mitigation.
Otherwise, we should implement cross-node traffic. We had a discussion about this way back, but the networking team never got around to implementing it. If it is required now we could try to tackle it.

@DockToFuture / @axel7born : FYI.

**Comment by acumino** [comment]
> Thanks for the detailed analysis! That's indeed quite helpful. Did you folks discuss next steps? What can we do now to solve or at least mitigate this problem?

One bandaid solution to remove the flakiness of the test is to remove the MR of `local-path-provisioner` from the health check.

And then fixing inter-node communication can be done as separate PR.

**Comment by rfranzke** [comment]
> One banded solution to remove the flakiness of the test is to remove the MR of local-path-provisioner from the health check.

How would this work? The shoot care controller is checking all `ManagedResource`s, and we don't want code like `if name == "local-path-provisioner" { continue }` in this just to make our local test setup work.

> Well, if it only about name resolution and it can be ensured that every node gets a coredns replica [internal traffic policy](https://kubernetes.io/docs/concepts/services-networking/service-traffic-policy/) can be a mitigation.

Sounds reasonable as a mitigation. `provider-local` could modify the CoreDNS configuration via shoot webhook to make this work. @acumino would you look into this?

> Otherwise, we should implement cross-node traffic. We had a discussion about this way back, but the networking team never got around to implementing it. If it is required now we could try to tackle it.

Yes, please. It was to be expected that this hits us sooner or later. It would be great if we could make this work.

**Comment by seshachalam-yv** [comment]
We can mitigate this problem by creating a pod called "datapath" that can contain containers as machines. This way, the inter-node communication problem is resolved instead of creating a pod for each machine.

WDYT @ScheererJ @rfranzke

**Comment by acumino** [comment]
> How would this work? The shoot care controller is checking all ManagedResources, and we don't want code like if name == "local-path-provisioner" { continue } in this just to make our local test setup work.

I was talking about this annotation `resources.gardener.cloud/skip-health-check` to annotate `local-path-provisioner`.

> Sounds reasonable as mitigation. provider-local could modify the CoreDNS configuration via shoot webhook to make this work. @acumino would you look into this?

Yes, I can check.

**Comment by rfranzke** [comment]
> We can mitigate this problem by creating a pod called "datapath" that can contain containers as machines. This way, the inter-node communication problem is resolved instead of creating a pod for each machine.

Thanks for joining the discussion @seshachalam-yv, but this sounds crazy 😄 How should this work with the MCM implementation - you can't modify existing pods and add or remove containers to/from it.
I think this comes with its own set of problems and skews reality w.r.t. machine handling even more (having pods in the same namespace like the control plane components already comes with different boundary conditions compared to regular cloud-based shoots).

> I was talking about this annotation `resources.gardener.cloud/skip-health-check` to annotate `local-path-provisioner`.

Good idea, I didn't think about this option. But perhaps the "more sustainable" bandaid would still be to adapt the CoreDNS configuration until the real cluster networking works.

**Comment by ScheererJ** [comment]
@seshachalam-yv Yes, you could also move all machines into a single pod. However, this would have implications if you want to add/remove nodes. I would assume getting the routing right would be easier, but feel free to prove me wrong :-) .

**Comment by ScheererJ** [comment]
@rfranzke You can add debug containers later on (see https://kubernetes.io/docs/tasks/debug/debug-application/debug-running-pod/#ephemeral-container-example), but this would likely make the solution even crazier ;-) ...

**Comment by seshachalam-yv** [comment]
> Thanks for joining the discussion @seshachalam-yv, but this sounds crazy 😄 How should this work with the MCM implementation - you can't modify existing pods and add or remove containers to/from it. I think this comes with its own set of problems and skews reality w.r.t. machine handling even more (having pods in the same namespace like the control plane components already comes with different boundary conditions compared to regular cloud-based shoots).

Yes, I agree it's difficult.
I'm not aware of how MCM works. 
But we can have `machine-provisioner` pod (DIND), based on MCM machine request add/delete, it should create/delete container inside `machine-provisioner` pod.
